odl/operator/fspace_ops.py

```diff
@@ -0,0 +1,412 @@
+
+        Parameters
+        ----------
+        space : `FunctionSpace`
+            The space of elements which the operator is acting on
+        voxel_size : `n`-tuple of `floats`
+            `n`-tuple of scalars specifying the sample distances in each
+            dimension. Default distance is 1 in each direction
+        edge_order : {1, 2}, optional
+            Partial derivative is calculated using Nth order accurate
+            differences at the boundaries. Default edge order: 2
+        zero_padding : `bool`, optional
+            Implicit zero padding. Assumes values outside of `space` to be
+            zero. Default: True
+        """
+        self.voxel_size = voxel_size
+        self.edge_order = edge_order
+        self.zero_padding = zero_padding
+        super().__init__(domain=space,
+                         range=ProductSpace(space, len(voxel_size)),
+                         linear=True)
+
+    def _apply(self, rhs, out):
+        """Apply gradient operator to `rhs` and store result in `out`.
+
+        Parameters
+        ----------
+        rhs : ``domain`` element
+            Input vector to which the `DiscreteGradient` operator is applied to
+        out : ``range`` element
+            Output vector to which the result is written
+
+        Examples
+        --------
+        >>> from odl import uniform_discr, FunctionSpace, IntervalProd
+        >>> def ndvolume(vol_size, ndim, dtype=None):
+        ...     s = [1]
+        ...     vol = np.arange(vol_size, dtype=dtype)
+        ...     for _ in range(ndim - 1):
+        ...         s.insert(0, vol_size)
+        ...         vol = vol * vol.reshape(s)
+        ...     return vol
+        >>> ndim = 3
+        >>> vsize = 100
+        >>> disc = uniform_discr(FunctionSpace(IntervalProd(
+        ... [0.]*ndim, [vsize]*ndim)), [vsize]*ndim)
+        >>> f = disc.element(ndvolume(vsize, ndim, np.int32))
+        >>> A = DiscreteGradient(disc, (1.,)*ndim, zero_padding=True)
+        >>> Af = A(f)
+        >>> g = A.range.one()
+        >>> Adg = A.adjoint(g)
+        >>> g.inner(Af) - f.inner(Adg)
+        0.0
+        >>> B = DiscreteDivergence(disc, (1.,)*ndim, zero_padding=True)
+        >>> Bg = B(g)
+        >>> Bdf = B.adjoint(f)
+        >>> f.inner(Bg) - g.inner(Bdf)
+        0.0
+        >>> ndim = 3
+        >>> vsize = 9
+        >>> disc = uniform_discr(FunctionSpace(IntervalProd(
+        ... [0.]*ndim, [vsize]*ndim)), [vsize]*ndim)
+        >>> f = disc.element(ndvolume(vsize, ndim, np.int32))
+        >>> A = DiscreteGradient(disc, (1.,)*ndim, zero_padding=True)
+        >>> Af = A(f)
+        >>> g = A.range.one()
+        >>> Adg = A.adjoint(g)
+        >>> g.inner(Af) - f.inner(Adg)
+        0.0
+        >>> B = DiscreteDivergence(disc, (1.,)*ndim, zero_padding=True)
+        >>> Bg = B(g)
+        >>> Bdf = B.adjoint(f)
+        >>> f.inner(Bg) - g.inner(Bdf)
+        0.0
+
+        """
+        rhs_data = np.asanyarray(rhs)
```

Comment:
Not really. That what used in the original numpy code for the gradient. I'll change it.